### PR TITLE
Preliminary SILGen support for address-only typed throws

### PR DESCRIPTION
**Pull request by slavapestov**
If the thrown error type is address-only, our function type has an indirect error result. Handle this in type lowering and SILGen prolog/epilog emission.

Next steps are introducing a zero-operand `throw_addr` instruction to exit out of a block after initializing the thrown error result, and setting up a stack-allocated box to receive the indirect error result from a try_apply, all the conversions and re-abstraction support, etc.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/SIL/IR/AbstractionPattern.cpp [hunk 1]
```diff
@@ -1953,16 +1995,57 @@ AbstractionPattern::getParameterConvention(TypeConverter &TC) const {
...
+  case Kind::ObjCCompletionHandlerArgumentsType:
+    llvm_unreachable("should not get here");
+  }
+}
+
+AbstractionPattern::CallingConventionKind
+AbstractionPattern::getErrorConvention(TypeConverter &TC) const {
+  // Tuples should be destructured.
+  if (isTuple()) {
+    return Destructured;
```
Comment (DougGregor) [inline lib/SIL/IR/AbstractionPattern.cpp:2019]: We ban tuple types from being thrown up in the type checker now, and even if we allowed them there... I think we'd still pass them indirectly here, right?

File: lib/SIL/IR/SILType.cpp [hunk 2]
```diff
@@ -713,6 +713,15 @@ bool SILModuleConventions::isPassedIndirectlyInSIL(SILType type, SILModule &M) {
   return false;
 }
 
+bool SILModuleConventions::isThrownIndirectlyInSIL(SILType type, SILModule &M) {
+  if (SILModuleConventions(M).loweredAddresses) {
```
Comment (DougGregor) [inline lib/SIL/IR/SILType.cpp:717]:
```suggestion
  if (SILModuleConventions(M).useLoweredAddresses()) {
```

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/SIL/IR/AbstractionPattern.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/SIL/IR/AbstractionPattern.cpp:2049]: Good catch, I’ll remove this. It was just a copy and paste

**Pull request merged**